Override canvas color from config identical to quality

ToBe998 opened the pull request.

rarila commented:
I see a logical problem, not only in your code, but also several times in the lines above your addition. I guess it either has to be
```php
if ($this->app['config']->get('foo/bar') !== null) {
    $val = $this->app['config']->get('foo/bar');
}
```
without default value (and no Yoda in the if clause :-) or just
```php
$val = $this->app['config']->get('foo/bar', 'Default value');
```
(assuming from code the second parameter if the get method is the default value if no value was set in config)

tbullacnd commented:
You are absolutly right. I just didnt want to change already approved code nor introduce a second code style in my very first addition to bolt.

If it was my code, i would refactor the whole section according to your second suggestion. Its not only more logical but also cleaner and easier to read.

Should I do the refactor?

CarsonF commented:
Keep in mind this is also the 2.x branch. If you want this functionality applied in 3.x it will need to be written for master branch as well (which is completely rewritten).

rarila commented:
@ToBe998 I would change it the lines of this PR with another commit here and would do the refactor of the other in another PR.

If suggestion one or two is right has to be checked in particular case. Suggestion one probably is the saver one as you could assume the variable to be set has already a default value set and only has to be overwritten if there's a value in config. Something like

```php
$this->resizer->canvas = $this->app['config']->get('general/thumbnails/canvas', $this->resizer->canvas);
```
would probably fit better but I think the version with the if clause is much more comprehensible although it is longer.

Also @CarsonF 's suggestion is to keep in mind.

ToBe998 commented:
I changed to code style again as we discussed. I finally settled on the "if variant" as it is closer to the current code style. I removed the defaults to fall back to the current variables.

The "if variant" is also kept as refactoring code to use the "one-line-variante" above is not applicable to all  parts of this section. Some parts have more logic which can not be done int his style (See top three config variables).

Btw, this is actually a rather small change. Might not be worth reimplementing large parts of this class. ;)

GawainLynch commented:
@CarsonF we should combine a forward port of this with merging.

rossriley commented:
I'm ok with this going in the 2.2 branch because it's a simple way to solve the problem.

But for 3.0 it would be nicer to not have it as a config setting but provide the option through the Twig helper, since it makes more sense being an option of the image render rather than being a global config.

GawainLynch commented:
Fair point… and 3.0 for both packages is a major bump.

Merge objections then?

ToBe998 commented:
Actually, I think all options should be available inside config.yml (probably also in theme/config.yml) and as a parameter to the twig function.

For example, a thumbnail of a car might have a different 404 image than a person's thumbnail. The quality of a thumbnail in a large teaser might have a higher quality setting than a tiny thumbnail in a navigation.

So my dream solution would be this:
All options should be available in config.yml, theme/config.yml AND as a parameter in a options object for twig. With a clear order of precedence.

```
<img src="{{thumbnail(record.image, 500, 300, "c", {"quality": 60, "canvas": [0,0,255]})"/>
```

This solution would be backwards compatible and could be enhanced with any yet unknown future parameter without trouble.

I'll try to get a proper new pull request with appropriate code in the next couple of days (if i have time) if you are interested. But let's stay agile for the moment and get this tiny feature done first, ok? ;)

GawainLynch commented:
Introducing a config setting into a stable branch is not a good idea, especially late in the game (branch life cycle), especially lacking documentation and forward direction.

Same deal… ideas? :grin:

ToBe998 commented:
Why is it a bad idea? It's optional and does not change already present functionality.

GawainLynch commented:
The bad idea is simply introducing a feature that is not present in a for forward branch… not that the feature itself is bad.

GawainLynch commented:
…well that an my lack of Engrish skillz :wink:

ToBe998 commented:
Ok, the missing feature in the forward branch is definetly a problem. Guess I'll have to checkout branch 3.x earlier than i thought. ;)
